ksp_plugin/part.cpp

```diff
@@ -132,43 +131,39 @@ void Part::ClearHistory() {
   history_ = prehistory_->NewForkAtLast();
 }
 
-void Part::set_containing_pile_up(IteratorOn<std::list<PileUp>> const pile_up) {
+void Part::set_containing_pile_up(
+    not_null<std::shared_ptr<PileUp>> const& pile_up) {
   CHECK(!is_piled_up());
   LOG(INFO) << "Adding part " << ShortDebugString() << " to the pile up at "
-            << &*pile_up.iterator();
+            << pile_up;
   containing_pile_up_ = pile_up;
 }
 
-std::experimental::optional<IteratorOn<std::list<PileUp>>>
-Part::containing_pile_up() const {
-  return containing_pile_up_;
+PileUp* Part::containing_pile_up() const {
+  return containing_pile_up_.get();
 }
 
 bool Part::is_piled_up() const {
-  // TODO(egg): |has_value()| once we have a standard |optional|.
-  return static_cast<bool>(containing_pile_up_);
+  return containing_pile_up_ != nullptr;
 }
 
-void Part::clear_pile_up() {
-  if (is_piled_up()) {
-    IteratorOn<std::list<PileUp>> pile_up = *containing_pile_up_;
-    for (not_null<Part*> const part : pile_up.iterator()->parts()) {
-      LOG(INFO) << "Removing part " << part->ShortDebugString()
-                << " from its pile up at " << &*pile_up.iterator();
-      part->containing_pile_up_ = std::experimental::nullopt;
-    }
-    CHECK(!is_piled_up());
-    pile_up.Erase();
-  }
+void Part::remove_from_pile_up() {
```

Comment:
The name is a bit misleading, in that the function does not remove `this` from `containing_pile_up_->parts_` (thus it makes the `PileUp` invalid.

`reset_containing_pile_up()` for consistency with the other two functions related to this member, or `reset_and_invalidate_containing_pile_up()` to make the invalidation explicit?

Reply:
Done.